expand docs (#371)
ZcashFoundation/frost: book/src/tutorial.md

# Tutorial

> The ZF FROST library consists of multiple crates. `frost-core` contains

Reviewer: ```suggestion
The ZF FROST suite consists of multiple crates. `frost-core` contains
```

a generic implementation of the protocol, which can't be used directly

> without a concrete suite.

Reviewer: ```suggestion
without a concrete instantiation.
```

`frost-core`, but also re-expose the `frost-core` functions without
generics. If you will only use a single ciphersuite, then we recommend
using those functions, and this tutorial will follow this approach.
If you need to support multiple ciphersuites then feel free to use
`frost-core` along with the ciphersuite types.

This tutorial will use the `frost-ristretto255` crate, but changing
to another ciphersuite should be a matter of simply changing the import.

## Including `frost-ristretto255`

Add to your `Cargo.toml` file:

```
[dependencies]
frost-ristretto255 = "0.3.0"
```

## Generating key shares with a trusted dealer

![Diagram of Trusted Dealer Key Generation, illustrating what is explained in the text](tutorial/tkg.png)

To generate the key shares, the dealer calls
[`generate_with_dealer()`](https://docs.rs/frost-ristretto255/latest/frost_ristretto255/keys/fn.generate_with_dealer.html).
It returns a `HashMap` mapping the (automatically generated) `Identifier`s to
their respective `SecretShare`s, and a `PublicKeyPackage` which contains the
`VerifyingShare` for each participant and the group public key (`VerifyingKey`).

TODO: insert code snippet

> Each `SecretShare` must then be sent via a **authenticated** and
> **confidential** channel for each participant, who verify the

Reviewer: ```suggestion
Each `SecretShare` must then be sent via an [**authenticated** and
**confidential** channel ](https://frost.zfnd.org/terminology.html#peer-to-peer-channel)for each participant, who verify the
```

> package to obtain a `KeyPackage` which contains their secret share,

Reviewer: ```suggestion
package to obtain a `KeyPackage` which contains their signing share,
```

verifying share and group public key.

TODO: insert code snippet

```admonish info
Currently there is no way to specify which identifiers to use. This will
likely be supported in the future.
```

```admonish info
Which encoding to use to transmit the `SecretShare` is up for the developer.
Note that specific types (Scalars, Elements) do have a standardized encoding.
Also note that `serde` support will be added in the future to offer a default
method for encoding.

The same applies to `KeyPackage`.
```

```admonish danger

> Which **authenticated** and **confidential** channel to use is up to the

Reviewer: ```suggestion
Which [**authenticated** and **confidential** channel](https://frost.zfnd.org/terminology.html#peer-to-peer-channel) to use is up to the
```

application. Some examples:

- Manually require the dealer to sent the `SecretShare`s to the
  partipants using some secure messenger such as Signal;
- Use a TLS connection, authenticating the server with a certificate
  and the client with some user/password or another suitable authentication
  mechanism;

> 

Reviewer: This duplicates a little bit of the peer-to-peer section in the terminology page, which I'm actually ok with as it's not easy to just inline that section fully in here, but we can link too

Author: Good point, you added a link but I also added a more explicit reference in 5d55082

> Failure of using a **confidential** channel may lead to the shares being
> stolen and possibly allowing signature forgeries if a threshold of them
> are stolen.

Reviewer: ```suggestion
Failure of using a **confidential** channel may lead to the shares being
stolen and possibly allowing signature forgeries if a threshold number of 
them are stolen.
```

Failure of using an **authenticated** channel may lead to shares being
sent to the wrong person, possibly allowing unintended parties
to generate signatures.
```

```admonish danger
The `SecretPackage` contents must be stored securely. For example:

- Make sure other users in the system can't read it;
- If possible, use the OS secure storage such that the package
  contents can only be opened with the user's password or biometrics.
```

```admonish warning

> The participants may wish to not fully trust the dealer. While **the dealer

Reviewer: ```suggestion
The participants may wish to not fully trust the dealer. **The dealer
```
